### DISCUSSION TITLE
Nuxt3 Logging with request id

### DISCUSSION THREAD

**Discussion by marc-arnoult** [nuxt/nuxt#27253] [Questions] [answered]
Hi,

In my current Nuxt 2 project, I use AsyncLocalStorage to set the request ID in the logger within the request context. This approach eliminates the need to pass the request ID throughout the application manually.

For this feature, I employ Express with a middleware to enable it. This middleware is the first server middleware in the stack.

Here is an example:

```
const loggerChild = process.winstonLog.child({
    request_id: requestId,
    correlation_id: correlationId,
    route: req.path,
});

asyncLocalStorage.run(loggerChild, () => {
    next();
});
```
With Nuxt 3, I no longer have this feature with serverMiddleware and the next() behavior.

Is it possible to achieve something similar in Nuxt 3? I've explored the asyncContext feature without success.

Regards,

Marc

**Comment by redsuperbat** [comment]
I was able to solve this by creating a server middleware in `server/middleware` which creates a request-id and attaches it to the event.context. 

`server/middleware/01-request-id.ts`

```ts
import { randomUUID } from "node:crypto";
export default defineEventHandler((event) => {
  event.context.requestId = randomUUID();
});
```

The event context can then be accessed in other route handlers/middlewares. 

`api/entity.get.ts`

```ts
export default defineEventHandler((event) => {
  console.log(event.context.requestId);
});
```



Hope it helps 🙏

**Comment by marc-arnoult** [comment, marked as answer]
Hi @redsuperbat thanks for you reply !
Indeed i did something like that, the problem was logging in other function / service / classes...

Without passing the context down to everyone.

I managed it with the experimental feature `asyncContext`.

i have a middleware, with this code ( i use Pino ) : 

```js
const child = logger.child({
   correlationId: correlationId,
});

event.context.logger = child;
```

Then in a service, i have something in that style : 
```js
export function log(level, ...args) {
  /**
   * https://nuxt.com/docs/guide/going-further/experimental-features#asynccontext
   */

    const event = useEvent();

    event.context.logger[level](args)
}
```
Now i can use it everywhere in the app without passing the `event` object.
